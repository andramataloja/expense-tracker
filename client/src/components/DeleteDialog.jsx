import React, { useState } from "react";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Box,
  makeStyles
} from "@material-ui/core";
import DeleteIcon from "@material-ui/icons/Delete";
import axios from "axios";

const useStyles = makeStyles(theme => ({
  deleteCommands: {
    color: theme.palette.secondary.dark
  }
}));

const DeleteDialog = props => {
  const classes = useStyles();
  const [open, setOpen] = useState(false);

  const handleDelete = () => {
    axios
      .delete("/delete/expense", {
        data: { expense_id: props.expense_id }
      })
      .then(res => console.log(res))
      .catch(err => console.log(err));
    window.location.reload(false);
  };

  return (
    <Box>
      <DeleteIcon color="secondary" onClick={() => setOpen(true)} />
      <Dialog
        open={open}
        onClose={() => setOpen(false)}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">{" Delete?"}</DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">
            Are you sure you want to delete this expense?
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button
            onClick={() => setOpen(false)}
            className={classes.deleteCommands}
          >
            Cancel
          </Button>
          <Button onClick={handleDelete} className={classes.deleteCommands}>
            Delete
          </Button>
        </DialogActions>
      </Dialog>
    </Box>
  );
};
export default DeleteDialog;
