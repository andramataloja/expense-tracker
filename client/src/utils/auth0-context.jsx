import React, { useState, useEffect, useContext } from "react";
import createAuth0Client from "@auth0/auth0-spa-js";

const DEFAULT_REDIRECT_CALLBACK = () =>
  window.history.replaceState({}, document.title, window.location.pathname);

export const Auth0Context = React.createContext();
export const useAuth0 = () => useContext(Auth0Context);
export const Auth0Provider = ({
  children,
  onRedirectCallback = DEFAULT_REDIRECT_CALLBACK
}) => {
  const [isAuthenticated, setIsAuthenticated] = useState();
  const [user, setUser] = useState();
  const [auth0Client, setAuth0Client] = useState();
  const [isLoading, setIsLoading] = useState(true);

  const config = {
    domain: process.env.REACT_APP_AUTH0_DOMAIN,
    client_id: process.env.REACT_APP_AUTH0_CLIENT_ID,
    redirect_uri: "http://localhost:3000/Dashboard"
  };

  useEffect(() => {
    const initializeAuth0 = async () => {
      const auth0FromHook = await createAuth0Client(config);
      setAuth0Client(auth0FromHook);

      if (
        window.location.search.includes("code=") &&
        window.location.search.includes("state=")
      ) {
        const { appState } = await auth0FromHook.handleRedirectCallback();
        onRedirectCallback(appState);
      }

      const isAuthenticated = await auth0FromHook.isAuthenticated();

      setIsAuthenticated(isAuthenticated);

      if (isAuthenticated) {
        const user = await auth0FromHook.getUser();
        setUser(user);
      }

      setIsLoading(false);
    };
    initializeAuth0();
  }, []);

  const handleRedirectCallback = async () => {
    setIsLoading(true);
    await auth0Client.handleRedirectCallback();
    const user = await auth0Client.getUser();
    setIsLoading(false);
    setIsAuthenticated(true);
    setUser(user);
  };
  return (
    <Auth0Context.Provider
      value={{
        isAuthenticated,
        user,
        isLoading,
        handleRedirectCallback,
        getIdTokenClaims: (...p) => auth0Client.getIdTokenClaims(...p),
        loginWithRedirect: (...p) => auth0Client.loginWithRedirect(...p),
        getTokenSilently: (...p) => auth0Client.getTokenSilently(...p),
        logout: (...p) => auth0Client.logout(...p)
      }}
    >
      {children}
    </Auth0Context.Provider>
  );
};
