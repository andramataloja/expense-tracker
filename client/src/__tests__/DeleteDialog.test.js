import React from "react";
import {
  cleanup,
  render,
  fireEvent,
  screen,
  wait
} from "@testing-library/react";
import DeleteDialog from "../components/DeleteDialog";
import { useAuth0 } from "../utils/auth0-context";
import { Provider } from "react-redux";
import { act } from "react-dom/test-utils";

import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import mockAxios from "axios";

const API = "http://localhost:3000";

const user = {
  email: "johndoe@me.com",
  email_verified: true,
  sub: "google-oauth2|12345678901234"
};

jest.mock("../utils/auth0-context");
jest.mock("../store.js");

describe("Add Expense", () => {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);

  const mockResponse = {
    data: { expense_id: 116 },
    status: 200,
    statusText: "OK",
    headers: {},
    config: {}
  };

  afterEach(cleanup);
  beforeEach(() => {
    useAuth0.mockReturnValue({
      isAuthenticated: true,
      user,
      logout: jest.fn(),
      loginWithRedirect: jest.fn()
    });
    require("mutationobserver-shim");
  });

  it("should take a snapshot of delete button", async () => {
    const store = mockStore();
    const { asFragment } = render(
      <Provider store={store}>
        <DeleteDialog />
      </Provider>
    );
    await act(wait);

    expect(asFragment(<DeleteDialog />)).toMatchSnapshot();
  });

  it("delete dialog renders", async () => {
    const store = mockStore();
    render(
      <Provider store={store}>
        <DeleteDialog />
      </Provider>
    );
    await act(wait);

    fireEvent.click(screen.getByTestId("delete-button"));
    expect(screen.getByText("Delete?")).toBeVisible();
    expect(
      screen.getByText("Are you sure you want to delete this expense?")
    ).toBeVisible();
    expect(screen.getByText("Cancel")).toBeVisible();
    expect(screen.getByText("Delete")).toBeVisible();
  });

  it("delete dialog opens and closes", async () => {
    const store = mockStore();
    render(
      <Provider store={store}>
        <DeleteDialog />
      </Provider>
    );
    await act(wait);

    expect(screen.queryByTestId("delete-dialog")).not.toBeInTheDocument();
    fireEvent.click(screen.getByTestId("delete-button"));
    expect(screen.getByTestId("delete-dialog")).toBeVisible();
    fireEvent.click(screen.getByText("Cancel"));
    await wait(() =>
      expect(screen.queryByTestId("delete-dialog")).not.toBeInTheDocument()
    );
  });

  it("deletes expense", async () => {
    mockAxios.delete.mockImplementationOnce(() =>
      Promise.resolve(mockResponse)
    );
    const store = mockStore();
    const handleDelete = jest.fn();
    render(
      <Provider store={store}>
        <DeleteDialog onClick={handleDelete} />
      </Provider>
    );
    await act(wait);
    fireEvent.click(screen.getByTestId("delete-button"));
    fireEvent.click(screen.getByText("Delete"));
    /*  await wait(() =>
      expect(mockAxios.delete).toHaveBeenCalledWith(`${API}/delete/expense`, {
        data: {
          expense_id: 116
        }
      })
    );
    expect(
      mockAxios.delete(`${API}/delete/expense`, {
        data: { expense_id: 116 }
      })
    ).resolves.toEqual(mockResponse); */

    await wait(() =>
      expect(screen.queryByTestId("delete-dialog")).not.toBeInTheDocument()
    );
  });
});
