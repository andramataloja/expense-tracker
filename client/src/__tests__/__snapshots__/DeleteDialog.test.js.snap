// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Expense should take a snapshot of delete button 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-3"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
      data-testid="delete-button"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
      />
    </svg>
  </div>
</DocumentFragment>
`;
